projects/params_vs_compute/hash_ladder/hash_ladder.py

```diff
@@ -0,0 +1,477 @@
+                            encoder_mask=encoder_mask,
+                            incr_state=incr_state.get(idx),
+                        )
+
+        return tensor, new_incr_state
+
+    def forward(self, input, encoder_state, incr_state=None):
+        """
+        Forward pass.
+
+        :param LongTensor[batch,seqlen] input:
+            The decoder inputs (partial or full decoded token IDs).
+        :param encoder_state:
+            Output from the encoder module forward pass.
+        :param incr_state:
+            The incremental state: a dictionary whose keys index the layers and whose
+            values contain the incremental state for each layer.
+        """
+        encoder_output, encoder_mask = encoder_state
+
+        seq_len = input.size(1)
+        positions = input.new(seq_len).long()
+        positions = torch.arange(seq_len, out=positions).unsqueeze(0)
+
+        if incr_state is not None:
+            # We're doing incremental decoding, so select only the most recent position
+            input = input[:, -1:]
+            if positions is not None:
+                positions = positions[:, -1:]
+        else:
+            incr_state = {}
+
+        tensor = self.forward_embedding(input, positions)
+
+        tensor = self.dropout(tensor)  # --dropout
+
+        tensor, new_incr_state = self.forward_layers(
+            tensor, encoder_output, encoder_mask, incr_state, original_input=input
+        )
+
+        if self.variant == 'prelayernorm':
+            tensor = _normalize(tensor, self.norm_embeddings)
+
+        return tensor, new_incr_state
+
+
+class HashLayer(TransformerDecoderLayer):
+    def __init__(
+        self,
+        opt: Opt,
+        n_heads: int,
+        embedding_size: int,
+        ffn_size: int,
+        attention_dropout: float = 0.0,
+        relu_dropout: float = 0.0,
+        dropout: float = 0.0,
+        activation: str = 'relu',
+        variant: str = 'aiayn',
+        **kwargs,
+    ):
+        super().__init__(opt, n_heads, embedding_size, ffn_size, **kwargs)
+        self.opt = opt
+        self.dim = embedding_size
+        self.ffn_dim = ffn_size
+        self.variant = variant
+        self.activation = activation
+        self.dropout = nn.Dropout(p=dropout)
+
+        self.self_attention = self.swappables.self_attention(
+            n_heads, embedding_size, dropout=attention_dropout
+        )  # type: ignore
+        self.norm1 = torch.nn.LayerNorm(embedding_size, eps=LAYER_NORM_EPS)
+
+        self.encoder_attention = self.swappables.encoder_attention(
+            n_heads, embedding_size, dropout=attention_dropout
+        )  # type: ignore
+        self.norm2 = torch.nn.LayerNorm(embedding_size, eps=LAYER_NORM_EPS)
+
+        self.ffn = HashLayerFFN(
+            opt,
+            embedding_size,
+            ffn_size,
+            relu_dropout=relu_dropout,
+            activation=activation,
+        )  # type: ignore
```

Comment:
Same here; it looks like we're only overloading `HashLayerFFN`? If so, you can just create your `TransformerDecoderLayer` like this without overriding:
```
TransformerDecoderLayer.with_components(feedforward=HashLayerFFN)
```

Reply:
ok, i'll look at these later after this initial merge, thanks

Reply:
(but i think i can't do that because i need to pass through the original inputs in the decoder layer)